Python usage example notebook [Unitary hack]

ahkatlio opened the pull request:
## Add Python Interface Example Notebook using JuliaCall
#89 
**Purpose:**

This PR introduces a new Jupyter Notebook example (`examples/Using_PauliPropagation_jl_from_Python.ipynb`) that demonstrates how to use `PauliPropagation.jl` from a Python environment. This addresses feedback regarding the desire for Python accessibility and showcases the use of `juliacall` as an interface.

**Motivation:**

As highlighted by users, providing a Python interface can significantly broaden the accessibility and usability of `PauliPropagation.jl`. This notebook serves as a practical guide for users who prefer or need to work within a Python ecosystem.

**Changes Included:**

*   **New Jupyter Notebook:** `examples/Using_PauliPropagation_jl_from_Python.ipynb`
    *   **Installation Instructions:** Guides users on installing `juliacall`.
    *   **Setup and Initialization:** Shows how to initialize Julia and import `PauliPropagation.jl` via `juliacall`.
    *   **Basic Example Replication:** Replicates the "Basic Example" from the `PauliPropagation.jl` documentation, demonstrating core functionalities like:
        *   Defining observables (`PauliString`).
        *   Creating circuit topologies and circuits (`bricklayertopology`, `tfitrottercircuit`).
        *   Setting parameters.
        *   Propagating observables (`propagate`).
        *   Calculating overlaps (`overlapwithzero`).
    *   **Visualization:** Includes an example of plotting the evolution of an observable's expectation value over time using `matplotlib`.
    *   **Data Type Handling:** Demonstrates conversion between Julia arrays and NumPy arrays.
    *   **Custom Julia Functions:** Shows how to define and call simple Julia functions from Python.
    *   **Accessing Documentation:** Illustrates how to fetch Julia documentation strings.
    *   **Advanced Example:** Provides a comparison of different truncation settings (`max_weight`) and visualizes their impact.

**How to Test/Run:**

1.  **Prerequisites:**
    *   Python 3.x installed.
    *   Jupyter Notebook or JupyterLab installed (`pip install notebook` or `pip install jupyterlab`).
    *   Julia installed and accessible in your PATH or configured for `juliacall`.
    *   The `juliacall` Python package installed (`pip install juliacall`).
    *   The `matplotlib` and `numpy` Python packages installed (`pip install matplotlib numpy`).
2.  **Steps:**
    *   Clone this branch.
    *   Navigate to the `examples/` directory.
    *   Launch Jupyter Notebook/Lab (e.g., `jupyter notebook` or `jupyter lab`).
    *   Open the `Using_PauliPropagation_jl_from_Python.ipynb` notebook.
    *   Run the cells sequentially.

**Key Features Demonstrated in the Notebook:**

*   Seamless integration of `PauliPropagation.jl` into a Python workflow.
*   Step-by-step execution of a typical simulation.
*   Practical examples of plotting and data manipulation across languages.

ahkatlio commented:
@MSRudolph  may you check this pr

MSRudolph commented:
Thanks @ahkatlio, this looks like a very nice notebook! 

I have a few comments/questions, some of them may just be because of inexperience with JuliaCall.
- Is it possible to have the `jl.PauliPropagation.funcs()` calls less verbose? Could we even import those functions into the global namespace? Or at least alias things more concicely.
- Is there a reason you are add PauliPropagation.jl via `Pkg.add(url="https://github.com/MSRudolph/PauliPropagation.jl")`? Does installation via `Pkg.add("PauliPropagation")` not work?
- I like the idea of plotting some evolution over time, but I am not sure the plot as a function of `dt` is that common. Could we try fixing a `dt` and plotting as a function of the layers? This can be done by looping over the layer index and calling something like `psum = propagate!(circuit_layer, psum, thetas_layer)` over and over, and evaluating `overlapwithzero(psum)` in-between.


Otherwise this already looks super useful!

ahkatlio commented:
Hello @MSRudolph,
Thank you for your feedback! I've addressed all three of your suggestions:

1. **Verbose Julia function calls**: I've added a dedicated section (2.1) for creating aliases that makes code more concise:
   - Added a module-level alias: `pp = jl.PauliPropagation`
   - Created function-level aliases in Julia's Main namespace (accessible via `jl.functionname`)
   - All subsequent code now uses these shorter aliases

2. **Package installation method**: I've updated the installation to use the General registry:
   ```python
   jl.seval("""
       using Pkg
       if !haskey(Pkg.project().dependencies, "PauliPropagation")
           Pkg.add("PauliPropagation")  # Install from registered package
       end
   """)
```
3. **Evolution over layers**: I've added two new sections (7 and 8) that implement this suggestion:
 - Section 7 shows evolution with fixed dt (0.1) across different numbers of layers
 - Section 8 compares different truncation settings on layer-wise evolution
 - Both provide physically intuitive visualizations of how the circuit depth affects results
 
 The notebook is now more concise, uses the standard installation method, and includes the layer-wise evolution visualization you suggested.

MSRudolph commented:
Thanks @ahkatlio, it's great to see things are so doable through Python.

I have a few more comments if you don't mind:
- I am really not a fan of the dt scan. Could you just stick with the one as a function of layers?
- You layered approach is quite inefficient. You don't need to run the full simulation again for every number of layers. Here is how it could be improved:
```julia
nq = 10
nl = 15

# just one layer at a time
layer = tfitrottercircuit(nq, 1)
thetas = fill(0.2, countparameters(layer));

pstr = PauliString(nq, :Z, 5)

psum = PauliSum(pstr)
expects = [overlapwithzero(psum)]
for l in 1:nl
    psum = propagate!(layer, psum, thetas)
    push!(expects, overlapwithzero(psum))
end
```
- Could we move to the `tiltedtfitrottercircuit()` which is not as trivial and thus less misleading to readers?
- And do you have an idea how we could simplify `jl.Symbol("Z")`? Could we define `Zsym = jl.Symbol("Z")`, including for `X` and `Y`? This should simplify writing circuits and Pauli strings/sums in Python.


Thanks for your hard work!

ahkatlio commented:
> If I am looking at this correctly, @ahkatlio, the final plots don't actually show the correct data. What you need is to use the in-place `propagate!()` function, or at least assign the output of the out-of-place `propagate()` function like `psum = propagate(...)`.

Hello @MSRudolph 
You were absolutely right about the `propagate` function! I've updated the notebook to correctly re-assign the output of `propagate` back to the `psum` (and `psum_trunc`) variables in the layer-wise evolution loops.

The plots should now accurately reflect the evolution of the observable as layers are applied.

Thank you for spotting that critical detail! Please let me know if you see anything else or if this looks good now.

MSRudolph commented:
Thanks a lot @ahkatlio. This looks good to me, and I have sent this to my collaborators for review. 

Could you change the PR to the `dev` branch?

ahkatlio commented:
Hello @MSRudolph, 
Just checking, is everything good?

MSRudolph commented:
Thanks for your patience @ahkatlio. I'm still collecting feedback by other members of my group, and this may take a while. Don't worry about the bounty - if this PR goes through, the timing doesn't matter too much.

That being said, one kind of feedback I already got is that it looks too much like pure Julia. For example: j`l.fill(dt, pp.countparameters(single_layer))` is a pretty niche Julia-esque way of creating parameters. Instead, Python users would want to use numpy arrays. Similarly, it would be nice to showcase how to build topologies from Python lists like `[(1, 2), (2, 5), ...]`
In short, we would like to create a notebook more from the eyes of a person who only knows Python, using Python as much as possible.

Another feedback I got is that the notebook displays in what is possible, but comparing for example to our most basic example notebook, it lacks educational hand-holding, explaining readers what the JuliaCall library does, what the objects are, and also what the limitations of this approach are. It appears a less pedagogical than we might like. 

Let me know what you think about this feedback.

ahkatlio commented:
Hello @MSRudolph 
Based on the changes made to address the feedback, here's what I would suggest for your PR comment:

Thank you for the excellent updates to the Python interface notebook! The changes have thoroughly addressed our feedback and significantly improved the notebook in several key areas:

1. **More Python-centric approach**: 
   - Using NumPy arrays for parameter creation (`numpy_layer_params = np.full(num_layer_params, dt)`)
   - Creating a helpful `create_topology_from_python()` function to work with Python data structures
   - Overall making the code feel more natural to Python users

2. **Enhanced educational content**:
   - Great introduction explaining what JuliaCall is and its benefits/challenges
   - Excellent "Behind the Scenes" explanation of what's happening during propagation
   - Very useful "Limitations and Considerations" section that helps users understand when this approach is appropriate

3. **Clear examples and visualizations**:
   - The topology visualization from Python lists is particularly helpful
   - The systematic truncation analysis provides valuable insights
   - The comparisons between different approaches give users a good understanding of tradeoffs

The notebook now provides a much more accessible entry point for Python users while still showcasing the power of PauliPropagation.jl. It strikes a good balance between being educational and practical.


I appreciate your responsiveness to the feedback and the thoroughness of your implementation. This will be very valuable for users who want to leverage our package from Python!

ahkatlio commented:
Hello @MSRudolph Let me know about the changes

MSRudolph commented:
Hey @ahkatlio, despite the notebook being somewhat useful, I will have to be blunt here: It is clear that you are using language models to do a lot if not everything. Both your replies (see the one just above where it is most obvious) and your example notebook are clearly generated. If you are still interested in contributing, I will give you a chance to change your notebook to a human-written version and reply in person.

ahkatlio commented:
> Hey @ahkatlio, despite the notebook being somewhat useful, I will have to be blunt here: It is clear that you are using language models to do a lot if not everything. Both your replies (see the one just above where it is most obvious) and your example notebook are clearly generated. If you are still interested in contributing, I will give you a chance to change your notebook to a human-written version and reply in person.

give me 5 hour I will fix it

MSRudolph commented:
@ahkatlio I am now playing with the notebook. Thanks for the changes.

A few questions and comments remain:
- Could you clarify into which environment `juliacall` gets installed? Because it seems it is not the kernel that the notebook is running in. Is there a way to install into an existing Jupyter kernel?
- `create_topology_from_python()`: I don't think this function is required. This code would already work and be entirely from Python: `pp.tfitrottercircuit(2, 1, topology=[(1, 2), (2, 3)])`. No need to convert the topology manually. We want to use as much play Python as possible. Please check if this is also applicable elsewhere.
- similar but inverse to the case above, this code `pp.countparameters(circuit)` already returns an `Int`, which means it does not need to be converted again. Same with `overlapwithzero()` already returning a float.
- `parameters_numpy` does not need to be converted to a Julia array. `juliacall` does that for us.
- `pauli_sum_result` can be tested with `len(pauli_sum_result)` and does not need the extra function. It seems `juliacall` translates Python's `len()` to Julia's `length()`.

We are almost there. Could you please check to make the code in the notebook as simple and Pythonic as possible?

ahkatlio commented:
Hello @MSRudolph I have changed the notebook according to your comments. Please check.

MSRudolph commented:
Thank you, @ahkatlio. This looks good. Would you mind if I committed to your branch to make some wording/comment changes? Just to be a bit more in line with the style of our other notebooks.

ahkatlio commented:
> Thank you, @ahkatlio. This looks good. Would you mind if I committed to your branch to make some wording/comment changes? Just to be a bit more in line with the style of our other notebooks.

Sure thank you

ahkatlio commented:
> Thank you, @ahkatlio. This looks good. Would you mind if I committed to your branch to make some wording/comment changes? Just to be a bit more in line with the style of our other notebooks.

Would you merge this after that?

ahkatlio commented:
> Sure, and you will receive the bounty then. The changes I was planning to make aim to partly make the text have a few less LLM artifacts (especially the bullet points). Before I do that, could you please try to edit part of the text where you speak of `I` and `I struggled with this initially` to be a bit more general? We should be good to go soon.

Done

MSRudolph commented:
Hey @ahkatlio, please check out the PR to your fork: https://github.com/ahkatlio/PauliPropagation.jl/pull/1

After this, the only thing I would ask for is to simplify the variable names of things like `custom_connections_py` to `topology`, in line with some of our other notebooks. See cell [12] for some names I gave variables. Prefered names are `circuit`, `circuit_layer`, `parameters`, `thetas`, `topology`, `pstr`, `psum`, and similar. Then we can squash and merge :rocket:

MSRudolph commented:
Looks good to me, @ahkatlio. Could you please rename it back to `PP-from-Python.ipynb`? That would be name I would give it in the end.

ahkatlio commented:
> Great, anything else you would like to change before merging?

Nah its good I think but do I nee this? `More below vvv`

MSRudolph commented:
> Nah its good I think but do I nee this? `More below vvv` the comment on the installation section?

Feel free to remove this or do a final clean-up if you like.

MSRudolph commented:
Add example notebook for using PauliPropagation.jl from Python with JuliaCall.
Authored by @ahkatlio
